docs/using-the-compiler.rst

```diff
@@ -0,0 +1,175 @@
+files reside, so things like ``import "/etc/passwd";`` only work if you add ``=/`` as a remapping.
+
+If there are multiple matches due to remappings, the one with the longest common prefix is selected.
+
+If your contracts use :ref:`libraries <libraries>`, you will notice that the bytecode contains substrings of the form ``__LibraryName______``. You can use ``solc`` as a linker meaning that it will insert the library addresses for you at those points:
+
+Either add ``--libraries "Math:0x12345678901234567890 Heap:0xabcdef0123456"`` to your command to provide an address for each library or store the string in a file (one library per line) and run ``solc`` using ``--libraries fileName``.
+
+If ``solc`` is called with the option ``--link``, all input files are interpreted to be unlinked binaries (hex-encoded) in the ``__LibraryName____``-format given above and are linked in-place (if the input is read from stdin, it is written to stdout). All options except ``--libraries`` are ignored (including ``-o``) in this case.
+
+
+*****************************************
+Standardized Input and Output Description
+*****************************************
+
+The compiler API expects a JSON formatted input and outputs the compilations result in a JSON formatted output.
+
+Comments are of course not permitted and used here only for explanatory purposes.
+
+Input Description
+-----------------
+
+QUESTION: How to specific file-reading callback? - probably not as part of json input
+
+The input description is language-specific and could change with each compiler version, but it
+should be backwards compatible if possible.
+
+.. code-block:: none
+
+    {
+      sources:
+      {
+        // the keys here are the "global" names of the source files, imports can use other files via remappings (see below)
+        "abc": "contract b{}", // specify source directly
+        // (axic) I think 'keccak' on its on is not enough. I would go perhaps with swarm: "0x12.." and ipfs: "Qma..." for simplicity
+        // (chriseth) Where the content is stored is a second component, but yes, we could give an indication there.
+        "def": {keccak: "0x123..."}, // source has to be retrieved by its hash
+        "ghi": {file: "/tmp/path/to/file.sol"}, // file on filesystem
+        // (axic) I'm inclined to think the source _must_ be provided in the JSON,
+
+        "dir/file.sol": "contract a {}"
+      },
+      settings:
+      {
+        remappings: [":g/dir"], // just as it used to be
+        // (axic) what is remapping doing exactly?
+        optimizer: {enabled: true, runs: 500},
+        // if given, only compiles this contract, can also be an array. If only a contract name is given, tries to find it if unique.
+        compilationTarget: "myFile.sol:MyContract",
+        // addresses of the libraries. If not all libraries are given here, it can result in unlinked objects whose output data is different
+        libraries: {
+          "def:MyLib": "0x123123..."
```

Comment:
This is changed into a hashmap similarly to the metadata:
```
libraries: {
  "def": {
    "MyLib": "0x123123..."
  }
}
```

The top level key is the the name of the source file where the library is defined.  If remapping are used, this source file should match the remapping target.  If this key is an empty string, that refers to a global level.

Note: the metadata documentation will need to be updated too.